android/src/main/java/com/mapbox/reactnativemapboxgl/RNMGLAnnotationOptionsFactory.java

```diff
@@ -0,0 +1,225 @@
+import com.mapbox.mapboxsdk.geometry.LatLng;
+import com.mapbox.mapboxsdk.maps.MapboxMap;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+
+class RNMGLMarkerOptions implements RNMGLAnnotationOptions {
+    protected MarkerOptions _options;
+
+    public RNMGLMarkerOptions(MarkerOptions options) {
+        _options = options;
+    }
+
+    @Override
+    public Annotation addToMap(MapboxMap map) {
+        return map.addMarker(_options);
+    }
+}
+
+class RNMGLPolylineOptions implements RNMGLAnnotationOptions {
+    protected PolylineOptions _options;
+
+    public RNMGLPolylineOptions(PolylineOptions options) {
+        _options = options;
+    }
+
+    @Override
+    public Annotation addToMap(MapboxMap map) {
+        return map.addPolyline(_options);
+    }
+}
+
+class RNMGLPolygonOptions implements RNMGLAnnotationOptions {
+    protected PolygonOptions _options;
+
+    public RNMGLPolygonOptions(PolygonOptions options) {
+        _options = options;
+    }
+
+    @Override
+    public Annotation addToMap(MapboxMap map) {
+        return map.addPolygon(_options);
+    }
+}
+
+public class RNMGLAnnotationOptionsFactory {
+
+    public static RNMGLAnnotationOptions annotationOptionsFromJS(ReadableMap annotation, Context context) {
+        String type = annotation.getString("type");
+
+        if (type.equals("point")) {
+            return markerOptionsFromJS(annotation, context);
+        } else if (type.equals("polyline")) {
+            return polylineOptionsFromJS(annotation);
+        } else if (type.equals("polygon")) {
+            return polygonOptionsFromJS(annotation);
+        }
+
+        return null;
+    }
+
+    static Drawable drawableFromDrawableName(Context context, String drawableName) {
+        int resID = context.getResources().getIdentifier(drawableName, "drawable", context.getApplicationContext().getPackageName());
+        return ContextCompat.getDrawable(context, resID);
+    }
+
+    static Drawable drawableFromUrl(Context context, String url) throws IOException {
+        // This doesn't currently work, as it throws NetworkOnMainThreadException
+        HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
+        connection.connect();
+        InputStream input = connection.getInputStream();
+
+        Bitmap bitmap = BitmapFactory.decodeStream(input);
+        return new BitmapDrawable(context.getResources(), bitmap);
+    }
+
+    static Map<String, Icon> iconCache = new HashMap();
+
+    static Icon iconFromPathAndSize(Context context, String path, int width, int height) throws IOException {
+        String cacheKey = path + "||" + width + "||" + height;
+        Icon icon = iconCache.get(cacheKey);
+        if (icon != null) { return icon; }
+
+        Drawable drawable;
+        if (path.startsWith("image!")) {
```

Comment:
I wonder if we should depreciate `image!` with 5.0.0. `image!` has not been part of rn for a long time now. Perhaps we should look into passing a required image through.